pkg/media/oggwriter/oggwriter.go

```diff
@@ -0,0 +1,248 @@
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+
+	"github.com/pion/rtp"
+	"github.com/pion/rtp/codecs"
+)
+
+const (
+	pageHeaderTypeContinuationOfStream = 0x00
+	pageHeaderTypeBeginningOfStream    = 0x02
+	pageHeaderTypeEndOfStream          = 0x04
+	defaultPreSkip                     = 3840 // 3840 recommanded in the RFC
+)
+
+// OggWriter is used to take RTP packets and write them to an OGG on disk
+type OggWriter struct {
+	stream                  io.Writer
+	fd                      *os.File
+	sampleRate              uint32
+	channelCount            uint16
+	serial                  uint32
+	pageIndex               uint32
+	checksumTable           *[256]uint32
+	previousGranulePosition uint64
+	previousTimestamp       uint32
+	lastPayloadSize         int
+}
+
+// New builds a new OGG Opus writer
+func New(fileName string, sampleRate uint32, channelCount uint16) (*OggWriter, error) {
+	f, err := os.Create(fileName)
+	if err != nil {
+		return nil, err
+	}
+	writer, err := NewWith(f, sampleRate, channelCount)
+	if err != nil {
+		return nil, err
+	}
+	writer.fd = f
+	return writer, nil
+}
+
+// NewWith initialize a new OGG Opus writer with an io.Writer output
+func NewWith(out io.Writer, sampleRate uint32, channelCount uint16) (*OggWriter, error) {
+	if out == nil {
+		return nil, fmt.Errorf("file not opened")
+	}
+
+	writer := &OggWriter{
+		stream:        out,
+		sampleRate:    sampleRate,
+		channelCount:  channelCount,
+		serial:        rand.Uint32(),
+		checksumTable: generateChecksumTable(),
+
+		// Timestamp and Granule MUST start from 1
+		// Only headers can have 0 values
+		previousTimestamp:       1,
+		previousGranulePosition: 1,
+	}
+	if err := writer.writeHeaders(); err != nil {
+		return nil, err
+	}
+
+	return writer, nil
+}
+
+/*
+    ref: https://tools.ietf.org/html/rfc7845.html
+    https://git.xiph.org/?p=opus-tools.git;a=blob;f=src/opus_header.c#l219
+
+       Page 0         Pages 1 ... n        Pages (n+1) ...
+    +------------+ +---+ +---+ ... +---+ +-----------+ +---------+ +--
+    |            | |   | |   |     |   | |           | |         | |
+    |+----------+| |+-----------------+| |+-------------------+ +-----
+    |||ID Header|| ||  Comment Header || ||Audio Data Packet 1| | ...
+    |+----------+| |+-----------------+| |+-------------------+ +-----
+    |            | |   | |   |     |   | |           | |         | |
+    +------------+ +---+ +---+ ... +---+ +-----------+ +---------+ +--
+    ^      ^                           ^
+    |      |                           |
+    |      |                           Mandatory Page Break
+    |      |
+    |      ID header is contained on a single page
+    |
+    'Beginning Of Stream'
+
+   Figure 1: Example Packet Organization for a Logical Ogg Opus Stream
+*/
+
+func (i *OggWriter) writeHeaders() error {
+	// ID Header
+	oggIDHeader := make([]byte, 19)
+
+	copy(oggIDHeader[0:], []byte("OpusHead"))                       // Magic Signature 'OpusHead'
```

Comment:
maybe it will be worthy to put all the magic signatures as constants? maybethe compiler already optimizes this as it is

Reply:
Fixed! I think it just good for readability, nice to have things be explicit